File: C - C++/printMatrixInSpiral/printMatrixInSpiral.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function for printing matrix in spiral
void print(vector<vector<int> > arr, int i, int j, int m, int n)
{
    // If i or j lies outside the matrix
    if (i >= m or j >= n)
        return;
 
    // Print First Row
    for (int p = i; p < n; p++)
        cout << arr[i][p] << " ";
 
    // Print Last Column
    for (int p = i + 1; p < m; p++)
        cout << arr[p][n - 1] << " ";
 
    // Print Last Row, if Last and
    // First Row are not same
    if ((m - 1) != i)
        for (int p = n - 2; p >= j; p--)
            cout << arr[m - 1][p] << " ";
 
    // Print First Column,  if Last and
    // First Column are not same
    if ((n - 1) != j)
        for (int p = m - 2; p > i; p--)
            cout << arr[p][j] << " ";
 
    print(arr, i + 1, j + 1, m - 1, n - 1);
}

int main()
{
 
    vector<vector<int> > a;
    int r,c;
    cout<<"Enter number of Rows:"<<"\t";
    cin>>r;
    cout<<"Enter Number of Columns"<<"\t";
    cin>>c;
    for(int i = 0;i<r;i++){
        vector<int> x;
        for(int j = 0;j<c;j++){
            int val;
            cin>>val;
            x.push_back(val);
        }
        a.push_back(x);
    }
    print(a, 0, 0, r, c);
    return 0;
}

File: Java/Print Matrix In Spiral Order/printMatrixInSpiral.java
import java.io.*;
import java.util.Scanner;

public class spiral {

    // Driver Code
    public static void main(String[] args) {

        int m, n;
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter rows and columns ");
        m = sc.nextInt(); //input number of rows
        n = sc.nextInt(); // input number of columns
        int a[][] = new int[m][n];

        // Taking matrix from the user
        System.out.println("Enter matrix elements:");
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++)
                a[i][j] = sc.nextInt();
        }

        int i, k = 0, l = 0;

        /*  k - starting row index
        m - ending row index
        l - starting column index
        n - ending column index
        i - iterator
        */

        while (k < m && l < n) {
            // Print the first row from the remaining rows
            for (i = l; i < n; ++i) {
                System.out.print(a[k][i] + " ");
            }
            k++;

            // Print the last column from the remaining
            // columns
            for (i = k; i < m; ++i) {
                System.out.print(a[i][n - 1] + " ");
            }
            n--;

            // Print the last row from the remaining rows */
            if (k < m) {
                for (i = n - 1; i >= l; --i) {
                    System.out.print(a[m - 1][i] + " ");
                }
                m--;
            }

            // Print the first column from the remaining
            // columns */
            if (l < n) {
                for (i = m - 1; i >= k; --i) {
                    System.out.print(a[i][l] + " ");
                }
                l++;
            }
        }

    }
}